src/blocking/spi.rs

```diff
@@ -108,3 +108,134 @@ pub mod write_iter {
         }
     }
 }
+
+/// Provides SpiWithCS wrapper around an spi::* and OutputPin impl
+pub mod spi_with_cs {
+
+    use core::marker::PhantomData;
+    use core::fmt::Debug;
+
+    use crate::digital::OutputPin;
+    use super::*;
+
+    /// SpiWithCS wraps an blocking::spi* implementation with Chip Select (CS)
+    /// pin management.
+    /// For sharing SPI between peripherals, see [shared-bus]()
+    pub struct SpiWithCs<Spi, SpiError, Pin, PinError> {
+        spi: Spi,
+        cs: Pin,
+
+        _spi_err: PhantomData<SpiError>,
+        _pin_err: PhantomData<PinError>,
+    }
+
+    /// SpiWithCsErr provies an error numberation over generic Spi and Pin variants
+    #[derive(Clone, Debug, PartialEq)]
+    pub enum SpiWithCsErr<SpiError, PinError> {
+        /// Underlying SPI error
+        Spi(SpiError),
+        /// Underlying Pin error
+        Pin(PinError),
+    }
+
+    /// ManagedCS marker trait indicates Chip Select management is automatic
+    impl <Spi, SpiError, Pin, PinError> ManagedCs for SpiWithCs <Spi, SpiError, Pin, PinError> {}
+    
+    impl <Spi, SpiError, Pin, PinError> SpiWithCs <Spi, SpiError, Pin, PinError> 
+    where
+        Pin: crate::digital::OutputPin<Error=PinError>,
+        SpiError: Debug,
+        PinError: Debug,
+    {
+        /// Create a new SpiWithCS wrapper with the provided Spi and Pin
+        pub fn new(spi: Spi, cs: Pin) -> Self {
+            Self{spi, cs, _spi_err: PhantomData, _pin_err: PhantomData}
+        }
+
+        /// Fetch references to the inner Spi and Pin types.
+        /// Note that using these directly will violate the `ManagedCs` constraint.
+        pub fn inner(&mut self) -> (&mut Spi, &mut Pin) {
+            (&mut self.spi, &mut self.cs)
+        }
+    }
+
+    impl <Spi, SpiError, Pin, PinError> Transfer<u8> for SpiWithCs<Spi, SpiError, Pin, PinError> 
```

Comment:
This default implementation should be in a submodule so that it is explicitly imported. Otherwise it would not be possible for a third-party to offer their own implementation of the traits as it would result in conflicting implementations. e.g. to handle `CS` assertion timing requirements.

Reply:
i am not sure what you mean by this? it is not an opt-in default like the other trait defaults but and entirely separate container with CS, so, reuse is not expected?

For additional timing constraints one would need create a new container type (like `SpiWithCs` but `SpiWithTimedCs` or whatever and implement the traits on the new object, afaik trying to re-use both the container object and traits from this module would run afoul of the orphan rules (both type and trait would be external).

Reply:
Ah, I see now that `ManagedCs` is not part of the `spi_with_cs` module. That is all I need. Sorry for the confusion.